pdg/src/main.rs

```diff
@@ -50,10 +51,104 @@ impl Display for ToPrint {
     }
 }
 
+pub struct Pdg {
+    pub events: Vec<Event>,
+    pub metadata: Metadata,
+    pub graphs: Graphs,
+}
+
+impl Pdg {
+    pub fn new(metadata_path: &Path, event_log_path: &Path) -> eyre::Result<Self> {
+        let events = read_event_log(event_log_path)?;
+        let metadata = read_metadata(metadata_path)?;
+        let graphs = construct_pdg(&events, &metadata);
+        Ok(Self {
+            events,
+            metadata,
+            graphs,
+        })
+    }
+
+    pub fn repr<'a>(&'a self, to_print: &'a [ToPrint]) -> PdgRepr<'a> {
+        PdgRepr {
+            pdg: self,
+            to_print,
+        }
+    }
+}
+
+pub struct PdgRepr<'a> {
+    pub pdg: &'a Pdg,
+    pub to_print: &'a [ToPrint],
+}
+
+impl Display for PdgRepr<'_> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let Self {
+            pdg:
+                Pdg {
+                    events,
+                    metadata,
+                    graphs,
+                },
+            to_print,
+        } = self;
+        let should_print = |e| to_print.contains(&e);
+
+        if should_print(ToPrint::Metadata) {
+            writeln!(f, "{metadata:#?}")?;
+        }
+
+        if should_print(ToPrint::Events) {
+            for event in events.iter() {
+                let mir_loc = metadata.get(event.mir_loc);
+                let kind = &event.kind;
+                writeln!(f, "{mir_loc:?} -> {kind:?}")?;
+            }
+        }
+
+        if should_print(ToPrint::LatestAssignments) {
+            for ((func_hash, local), p) in &graphs.latest_assignment {
+                let func = &metadata.functions[func_hash];
+                writeln!(f, "({func}:{local:?}) => {p:?}")?;
+            }
+        }
+
+        for graph in &graphs.graphs {
+            let needs_write = graph
+                .needs_write_permission()
+                .map(|node_id| node_id.as_usize())
+                .collect::<Vec<_>>();
+            if should_print(ToPrint::Graphs) {
+                writeln!(f, "{graph}")?;
+            }
+            if should_print(ToPrint::WritePermissions) {
+                writeln!(f, "nodes_that_need_write = {needs_write:?}")?;
+            }
+            if should_print(ToPrint::Graphs) || should_print(ToPrint::WritePermissions) {
+                writeln!(f)?;
+            }
```

Comment:
I would move this `if` to surround the entire `for graph...` loop, and then make this `writeln!` unconditional. That would clarify that the `writeln!(f)` terminates each graph printed and would avoid iterating if we don't need to look at graphs.

Reply:
I'd like to change this in a separate PR, since this right now is just moving code, not changing anything.

Reply:
Wait, this is not quite equivalent.  If `--print write-permissions` is specified, but no `--print graphs`, the write permissions and no graphs should be printed.

Reply:
A separate PR for these changes is fine.

I think what I requested is equivalent, though--the three `writeln!`s in the loop are guarded by `should_print(ToPrint::Graphs)`, `should_print(ToPrint::WritePermissions)`, and their disjunction, respectively. Making the whole loop conditional on the disjunction doesn't impact the first two conditionals as they're refinements of it, but it makes the third fully redundant, so it can be made unconditonal:
```rust
        if should_print(ToPrint::Graphs) || should_print(ToPrint::WritePermissions) {
            for graph in &graphs.graphs {
                let needs_write = graph
                    .needs_write_permission()
                    .map(|node_id| node_id.as_usize())
                    .collect::<Vec<_>>();
                if should_print(ToPrint::Graphs) {
                    writeln!(f, "{graph}")?;
                }
                if should_print(ToPrint::WritePermissions) {
                    writeln!(f, "nodes_that_need_write = {needs_write:?}")?;
                }
                writeln!(f)?;
            }
        }
```

Reply:
Oh like that you mean.  I'm not sure that's much of an advantage, since it's just hosting the `if` around the `writeln!(f)` to outside the loop.

Reply:
The concrete advantage is that it skips iterating over the graphs if nothing about them needs to be printed. It's not a huge deal, but is a small "pay for what you use".

Reply:
Fixed in #645.